sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py

```diff
@@ -793,18 +814,28 @@ async def get_throughput(self, **kwargs: Any) -> ThroughputProperties:
         :rtype: ~azure.cosmos.offer.ThroughputProperties
         """
         response_hook = kwargs.pop('response_hook', None)
-        properties = await self._get_properties()
-        link = properties["_self"]
-        query_spec = {
-            "query": "SELECT * FROM root r WHERE r.resource=@link",
-            "parameters": [{"name": "@link", "value": link}],
-        }
-        throughput_properties = [throughput async for throughput in
-                                 self.client_connection.QueryOffers(query_spec, **kwargs)]
-        if len(throughput_properties) == 0:
-            raise CosmosResourceNotFoundError(
-                status_code=StatusCodes.NOT_FOUND,
-                message="Could not find ThroughputProperties for container " + self.container_link)
+        throughput_properties: List[Dict[str, Any]]
+        # In case of a container recreate we will run this a second time
+        for attempt in range(2):
+            if attempt == 1:
+                # Refresh cache if no offer throughput was received the first time
+                await self.read()
```

Comment:
good question, I am also wondering why this logic would be here as opposed to being included in the general retry policy we would have? does the get_throughput operation run with a wrong collection rid?

Reply:
Read offer behaves differently to the other operations as instead of erroring out with a 404 it returns an empty query/list/none so that is why the 404 error is raised in the logic here. It also has some preprocessing logic before the request is made that uses data from the container properties cache, so we retry doing it after refreshing the cache so it tries with the current containers properties instead of a stale one, then if it fails or succeeds it will be based on non stale data.
The read() also refreshes the container cache.